src/bun.js/api/bun/process.zig

```diff
@@ -0,0 +1,793 @@
+    OtherOperationCount: u64 = 0,
+    ReadTransferCount: u64 = 0,
+    WriteTransferCount: u64 = 0,
+    OtherTransferCount: u64 = 0,
+};
+
+extern "kernel32" fn GetProcessIoCounters(handle: std.os.windows.HANDLE, counters: *IO_COUNTERS) callconv(std.os.windows.WINAPI) c_int;
+
+pub fn uv_getrusage(process: *uv.uv_process_t) win_rusage {
+    var usage_info: Rusage = .{ .utime = .{}, .stime = .{} };
+    const process_pid: *anyopaque = process.process_handle;
+    const WinTime = std.os.windows.FILETIME;
+    var starttime: WinTime = undefined;
+    var exittime: WinTime = undefined;
+    var kerneltime: WinTime = undefined;
+    var usertime: WinTime = undefined;
+    // We at least get process times
+    if (std.os.windows.kernel32.GetProcessTimes(process_pid, &starttime, &exittime, &kerneltime, &usertime) == 1) {
+        var temp: u64 = (@as(u64, kerneltime.dwHighDateTime) << 32) | kerneltime.dwLowDateTime;
+        if (temp > 0) {
+            usage_info.stime.tv_sec = @intCast(temp / 10000000);
+            usage_info.stime.tv_usec = @intCast(temp % 1000000);
+        }
+        temp = (@as(u64, usertime.dwHighDateTime) << 32) | usertime.dwLowDateTime;
+        if (temp > 0) {
+            usage_info.utime.tv_sec = @intCast(temp / 10000000);
+            usage_info.utime.tv_usec = @intCast(temp % 1000000);
+        }
+    }
+    var counters: IO_COUNTERS = .{};
+    _ = GetProcessIoCounters(process_pid, &counters);
+    usage_info.inblock = counters.ReadOperationCount;
+    usage_info.oublock = counters.WriteOperationCount;
+
+    const memory = std.os.windows.GetProcessMemoryInfo(process_pid) catch return usage_info;
+    usage_info.maxrss = memory.PeakWorkingSetSize / 1024;
+
+    return usage_info;
+}
+pub const Rusage = if (Environment.isWindows) win_rusage else std.os.rusage;
+
+const Subprocess = JSC.Subprocess;
+const LifecycleScriptSubprocess = bun.install.LifecycleScriptSubprocess;
+const ShellSubprocess = bun.shell.ShellSubprocess;
+const ShellSubprocessMini = bun.shell.ShellSubprocessMini;
+pub const ProcessExitHandler = struct {
+    ptr: TaggedPointer = TaggedPointer.Null,
+
+    pub const TaggedPointer = bun.TaggedPointerUnion(.{
+        Subprocess,
+        LifecycleScriptSubprocess,
+        ShellSubprocess,
+        ShellSubprocessMini,
+    });
+
+    pub fn init(this: *ProcessExitHandler, ptr: anytype) void {
+        this.ptr = TaggedPointer.init(ptr);
+    }
+
+    pub fn call(this: *const ProcessExitHandler, comptime ProcessType: type, process: *ProcessType, status: Status, rusage: *const Rusage) void {
+        if (this.ptr.isNull()) {
+            return;
+        }
+
+        switch (this.ptr.tag()) {
+            .Subprocess => {
+                if (comptime ProcessType != Process)
+                    unreachable;
+                const subprocess = this.ptr.as(Subprocess);
+                subprocess.onProcessExit(process, status, rusage);
+            },
+            .LifecycleScriptSubprocess => {
```

Comment:
it would be interesting if this type of tagged pointer switch could be comptime-ified. i dont think we should do it. but it is something i think about

inline else => |t| { t.as(...).onProcessExit(...) }